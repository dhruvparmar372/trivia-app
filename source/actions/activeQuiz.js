export const startQuiz = (quiz, startedOn) => ({
  type: "QUIZ/START",
  quiz,
  startedOn
});

export const endQuiz = (quiz, endedOn) => ({
  type: "QUIZ/END",
  quiz,
  endedOn
});

export const recordAnswer = (quizId, questionId, answer) => ({
  type: "QUIZ/RECORD_ANSWER",
  quizId,
  questionId,
  answer
});
