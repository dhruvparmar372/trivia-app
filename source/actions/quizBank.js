import { createRoutine } from "redux-saga-routines";

export const fetchQuiz = createRoutine("QUIZBANK/FETCH");
export const addQuiz = quiz => ({
  type: "QUIZBANK/ADD",
  quiz
});

export const fillQuizBank = quizCount => ({
  type: "QUIZBANK/FILL",
  quizCount
});

export const setQuizFetchError = fetchErrorText => ({
  type: "QUIZBANK/SET_FETCH_ERROR",
  fetchErrorText
});
