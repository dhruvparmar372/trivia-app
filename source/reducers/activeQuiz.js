import {
  startQuiz as startQuizAction,
  endQuiz as endQuizAction,
  recordAnswer as recordAnswerAction
} from "source/actions/activeQuiz";

export const initialState = null;

export default function(state = initialState, action) {
  switch (action.type) {
    case startQuizAction().type:
      return {
        ...action.quiz,
        startedOn: action.startedOn
      };

    case endQuizAction().type:
      return initialState;

    case recordAnswerAction().type:
      const { quizId, questionId, answer } = action;
      return state && state.id === action.quizId && answer
        ? {
            ...state,
            questions: state.questions.map(question => {
              return question.id === questionId
                ? {
                    ...question,
                    answer
                  }
                : question;
            })
          }
        : state;

    default:
      return state;
  }
}
