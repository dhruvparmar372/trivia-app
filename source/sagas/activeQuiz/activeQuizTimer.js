import { delay } from "redux-saga";
import { take, fork, put, cancel } from "redux-saga/effects";
import {
  startQuiz as startQuizAction,
  endQuiz as endQuizAction,
  updateQuizTime as updateQuizTimeAction
} from "source/actions/activeQuiz";

let timerPaused = false;
let secondsElapsed = 0;
function* trackQuizTime() {
  while (true) {
    yield delay(1000);
    if (!timerPaused) {
      secondsElapsed += 1;
      yield put(updateQuizTimeAction(secondsElapsed));
    }
  }
}

export default function*() {
  yield take(startQuizAction().type);
  const timer = yield fork(trackQuizTime);

  yield take(endQuizAction().type);
  yield cancel(timer);
}
