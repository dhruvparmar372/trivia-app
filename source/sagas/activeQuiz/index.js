import { all, fork } from "redux-saga/effects";
import activeQuizTimer from "./activeQuizTimer";

export default function*() {
  yield all([activeQuizTimer].map(fork));
}
