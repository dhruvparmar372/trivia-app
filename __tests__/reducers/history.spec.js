import { endQuiz as endQuizAction } from "source/actions/activeQuiz";
import historyReducer, { initialState } from "source/reducers/history";

const quizStartDate = "2017-12-30T18:30:00.000Z";
const quizEndDate = "2017-12-31T18:30:00.000Z";
const dummyQuiz = {
  id: 1,
  startedOn: quizStartDate,
  questions: [
    {
      id: 1,
      type: "boolean",
      question: "1 + 1 = 2?",
      correct_answer: "True",
      incorrect_answers: ["False"]
    }
  ]
};

describe("history reducer", () => {
  test("saves quiz in data on endQuiz action", () => {
    expect(
      historyReducer(
        initialState,
        endQuizAction(
          {
            ...dummyQuiz,
            questions: [
              {
                id: 1,
                type: "boolean",
                question: "1 + 1 = 2?",
                correct_answer: "True",
                incorrect_answers: ["False"],
                answer: { value: "True" }
              }
            ]
          },
          quizEndDate
        )
      )
    ).toMatchSnapshot();
  });

  test("ignores saving quiz if not complete on endQuiz action", () => {
    expect(
      historyReducer(initialState, endQuizAction(dummyQuiz, quizEndDate))
    ).toMatchSnapshot();
  });
});
