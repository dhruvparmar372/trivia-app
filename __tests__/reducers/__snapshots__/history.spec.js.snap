// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`history reducer ignores saving quiz if not complete on endQuiz action 1`] = `
Object {
  "data": Array [],
}
`;

exports[`history reducer saves quiz in data on endQuiz action 1`] = `
Object {
  "data": Array [
    Object {
      "endedOn": "2017-12-31T18:30:00.000Z",
      "id": 1,
      "questions": Array [
        Object {
          "answer": Object {
            "value": "True",
          },
          "correct_answer": "True",
          "id": 1,
          "incorrect_answers": Array [
            "False",
          ],
          "question": "1 + 1 = 2?",
          "type": "boolean",
        },
      ],
      "startedOn": "2017-12-30T18:30:00.000Z",
    },
  ],
}
`;
